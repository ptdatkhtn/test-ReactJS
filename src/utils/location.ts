export const degreeToRadian = (degree : number): number => {
  return degree * Math.PI / 180
}

export const calculateDistanceInKm = (lat1: number, lon1: number, lat2: number, lon2: number): number => {
  const earthRadiusKm = 6371

  // have to convert degree unit to radian unit before the calculation
  const dLat = degreeToRadian(lat2 - lat1)
  const dLon = degreeToRadian(lon2 - lon1)

  lat1 = degreeToRadian(lat1)
  lat2 = degreeToRadian(lat2)

  const a = Math.sin(dLat / 2) * Math.sin(dLat / 2) + Math.sin(dLon / 2) * Math.sin(dLon / 2) * Math.cos(lat1) * Math.cos(lat2)
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))
  return earthRadiusKm * c
}