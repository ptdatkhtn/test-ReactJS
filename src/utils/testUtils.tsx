import { render as rtlRender } from '@testing-library/react'
import React from 'react'
import { ThemeProvider } from 'styled-components'
import theme from 'themes'

const render = (ui: JSX.Element, { ...renderOptions } = {}) => {
  const Wrapper = ({ children }: { children: React.ReactNode }) => {
    return (
      <ThemeProvider theme={theme}>
        {children}
      </ThemeProvider>
    )
  }
  return rtlRender(ui, { wrapper: Wrapper as React.FC, ...renderOptions })
}

const fixMatchMedia = () => {
  window.matchMedia = window.matchMedia || function() {
    return {
        matches: false,
        addListener: function() {},
        removeListener: function() {}
    }
  }
}

export * from '@testing-library/react'
// override render method from @testing-library/react
export { render, fixMatchMedia }
