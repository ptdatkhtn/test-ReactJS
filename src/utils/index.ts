import { calculateDistanceInKm } from './location'

export { calculateDistanceInKm }