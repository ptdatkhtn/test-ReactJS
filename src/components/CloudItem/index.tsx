import React from 'react'

import { Container, CloudName, CloudDescription } from './styles'
import { ICloudItemComp } from 'interface'

const CloudItem = ({
  key,
  value,
  description,
  name,
  onClick,
  ...otherProps
}: ICloudItemComp) => {
  const onClickHandler = () => {
    onClick(value)
  }

  return (
    <Container onClick={onClickHandler} {...otherProps}>
      <CloudName>{name}</CloudName>
      <CloudDescription>{description}</CloudDescription>
    </Container>
  )
}

export default CloudItem