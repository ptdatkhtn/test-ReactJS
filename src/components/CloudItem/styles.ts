import styled from 'styled-components'
import theme from 'themes'
import { List } from 'components'

const { Item } = List

export const Container = styled(Item)`
  && {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    cursor: pointer;
    padding: ${theme.spacing.xs};
    margin-bottom: ${theme.spacing.xs};
    border-radius: 0.5rem;
    box-shadow: 0 3px 8px 0 rgb(0 0 0 / 3%);
    background-color: white;
    border: 1px solid transparent;
    transition: border 0.2s linear;
    :hover {
      border: 1px solid ${theme.color.grey2};
    }
  }
`
export const CloudName = styled.span`
  color: ${theme.color.primary1};
  font-size: ${theme.font.m};
`
export const CloudDescription = styled.span`
  color: ${theme.color.grey1};
  font-size: ${theme.font.xs};
`
