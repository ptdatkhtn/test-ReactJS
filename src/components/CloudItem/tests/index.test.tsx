import React from 'react'
import 'jest-styled-components'
import CloudItem from '..'
import { render, fireEvent, screen } from 'utils/testUtils'

describe('Cloud Item', () => {
  it('should match snapshot and trigger click function', () => {
    const mockOnClick = jest.fn()
    const { container } = render(
      <>
        <CloudItem data-testid='cloud_item' value='amazon' description='cloud description' name='cloud name' onClick={mockOnClick}/>
      </>
    )
    
    expect(container).toMatchSnapshot()
    expect(mockOnClick).toHaveBeenCalledTimes(0)
    fireEvent.click(screen.getByTestId('cloud_item'))
    expect(mockOnClick).toHaveBeenCalledTimes(1)
  })
})