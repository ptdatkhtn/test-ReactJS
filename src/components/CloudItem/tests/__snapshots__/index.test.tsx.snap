// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cloud Item should match snapshot and trigger click function 1`] = `
.c0.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
  padding: 0.8rem;
  margin-bottom: 0.8rem;
  border-radius: 0.5rem;
  box-shadow: 0 3px 8px 0 rgb(0 0 0 / 3%);
  background-color: white;
  border: 1px solid transparent;
  -webkit-transition: border 0.2s linear;
  transition: border 0.2s linear;
}

.c0.c0:hover {
  border: 1px solid #8f919d;
}

.c1 {
  color: #ff514d;
  font-size: 2rem;
}

.c2 {
  color: #565867;
  font-size: 1.6rem;
}

<div>
  <li
    class="ant-list-item c0"
    data-testid="cloud_item"
  >
    <span
      class="c1"
    >
      cloud name
    </span>
    <span
      class="c2"
    >
      cloud description
    </span>
  </li>
</div>
`;
